subprojects/core/src/main/java/org/gradle/initialization/ProjectPropertySettingBuildLoader.java

```diff
@@ -22,139 +22,102 @@
-        private final Map<Pair<String, ? extends Class<?>>, Optional<PropertyMutator>> mutators = new HashMap<>();
-        private ImmutableMap.Builder<String, Object> extraProjectProperties;
-
-        CachingPropertyApplicator(Class<? extends Project> projectClass) {
-            this.projectClass = projectClass;
+        for (Map.Entry<String, String> entry : mergedProjectProperties.entrySet()) {
+            String propertyName = entry.getKey();
+            // This is intentional relaxation of the type to support an edge-case of GradleBuild task
+            // that allows passing non-String properties via `startParameter.projectProperties`.
+            // The latter map is typed `Map<String, String>` but due to type erasure, the actual values were never explicitly checked.
+            Object propertyValue = entry.getValue();
+            assignOrCollectProperty(project, extraProjectPropertiesBuilder, propertyName, propertyValue);
         }
 
-        void configureProperty(Project project, String name, Object value) {
-            if (isPossibleProperty(name)) {
-                assert project.getClass() == projectClass;
-                PropertyMutator propertyMutator = propertyMutatorFor(name, typeOf(value));
-                if (propertyMutator != null) {
-                    propertyMutator.setValue(project, value);
-                } else {
-                    extraProjectProperties.put(name, value);
-                }
-            }
-        }
+        installProjectExtraPropertiesDefaults(project, extraProjectPropertiesBuilder.build());
+    }
 
-        @Nullable
-        private static Class<?> typeOf(@Nullable Object value) {
-            return value == null ? null : value.getClass();
+    @SuppressWarnings("deprecation")
+    private static void assignOrCollectProperty(
+        Project project,
+        ImmutableMap.Builder<String, Object> extraProjectProperties,
+        String propertyName,
+        Object propertyValue
+    ) {
+        if (propertyName.isEmpty()) {
+            // Historically, we filtered out properties with empty names here.
+            // They could appear in case the properties file has lines containing only '=' or ':'
+            return;
         }
 
-        @Nullable
-        private PropertyMutator propertyMutatorFor(String propertyName, @Nullable Class<?> valueType) {
-            return mutators.computeIfAbsent(
-                Pair.of(propertyName, valueType),
-                key -> {
-                    assert key.left != null;
-                    return Optional.ofNullable(writeablePropertyIfExists(projectClass, key.left, key.right));
+        switch (propertyName) {
+            case "version":
+                project.setVersion(propertyValue);
+                break;
+            case "group":
+                project.setGroup(propertyValue);
+                break;
+            case "description":
+                if (propertyValue instanceof String) {
```

Comment:
👍 Perhaps keep a note that this `if` can be removed once Gradle no longer accepts non-String properties in Gradle 10